### PR TITLE
Fix Prow auto-scaling

### PR DESCRIPTION
**Pull request by coryrc**
Autoscaling Prow clusters intermittently fail with messages like
`Could not resolve host: github.com`. Retry clone when this is detected.

Fixes: https://github.com/knative/test-infra/issues/1082
Fixes: https://github.com/kyma-project/test-infra/issues/1198

### CODE REVIEW

**Comment by cjwagner** [comment]
```
ERROR: /home/prow/go/src/github.com/kubernetes/test-infra/prow/clonerefs/BUILD.bazel:3:1: GoCompilePkg prow/clonerefs/linux_amd64_pure_stripped/go_default_library%/k8s.io/test-infra/prow/clonerefs.a failed (Exit 1)
/b/f/w/prow/clonerefs/run.go:93:19: undefined: time
/b/f/w/prow/clonerefs/run.go:95:8: undefined: time
```
/ok-to-test

File: prow/clonerefs/run.go [hunk 1]
```diff
@@ -81,7 +82,24 @@ func (o Options) Run() error {
...
 			defer wg.Done()
 			for ref := range input {
-				output <- cloneFunc(ref, o.SrcRoot, o.GitUserName, o.GitUserEmail, o.CookiePath, env, oauthToken)
+				// Attempt to clone the repo
+				// If the cluster is autoscaling, it seems Prow can hit DNS too hard on new nodes
+				// So we retry if have a failure we can guess is due to DNS
+				// Exponential backoff is targeted for a little over 5 minutes on the last try
+				for i := 1; i < 4; i++ {
+					record := cloneFunc(ref, o.SrcRoot, o.GitUserName, o.GitUserEmail, o.CookiePath, env, oauthToken)
+					if record.Failed {
```
Comment (fejta) [inline prow/clonerefs/run.go:91]: Seems like the following would be simpler:
```golang
if !record.Failed {
  output <- record
  break
}
// sleep

File: prow/clonerefs/run.go [hunk 2]
```diff
@@ -81,7 +82,24 @@ func (o Options) Run() error {
 		go func() {
 			defer wg.Done()
 			for ref := range input {
-				output <- cloneFunc(ref, o.SrcRoot, o.GitUserName, o.GitUserEmail, o.CookiePath, env, oauthToken)
```
Comment (fejta) [inline prow/clonerefs/run.go:84]: This always writes to the output channel. The changes need to keep doing that.

File: prow/clonerefs/run.go [hunk 3]
```diff
@@ -81,7 +82,24 @@ func (o Options) Run() error {
...
+				// So we retry if have a failure we can guess is due to DNS
+				// Exponential backoff is targeted for a little over 5 minutes on the last try
+				for i := 1; i < 4; i++ {
+					record := cloneFunc(ref, o.SrcRoot, o.GitUserName, o.GitUserEmail, o.CookiePath, env, oauthToken)
+					if record.Failed {
+						failed := record.Commands[len(record.Commands)-1]
+						if strings.Contains(failed.Error, "Could not resolve host") {
+							seconds := time.Second * 25 * i * i
+							logrus.Infof("DNS failure detected; retrying clone in %d seconds", seconds)
+							time.Sleep(seconds)
```
Comment (fejta) [inline prow/clonerefs/run.go:96]: We should not sleep when `i == 3`

File: prow/clonerefs/run.go [hunk 4]
```diff
@@ -81,7 +82,24 @@ func (o Options) Run() error {
...
+				// Attempt to clone the repo
+				// If the cluster is autoscaling, it seems Prow can hit DNS too hard on new nodes
+				// So we retry if have a failure we can guess is due to DNS
+				// Exponential backoff is targeted for a little over 5 minutes on the last try
+				for i := 1; i < 4; i++ {
+					record := cloneFunc(ref, o.SrcRoot, o.GitUserName, o.GitUserEmail, o.CookiePath, env, oauthToken)
+					if record.Failed {
+						failed := record.Commands[len(record.Commands)-1]
+						if strings.Contains(failed.Error, "Could not resolve host") {
+							seconds := time.Second * 25 * i * i
```
Comment (fejta) [inline prow/clonerefs/run.go:94]: `delay` or something along those lines is a better name (the int64 in this variable is not 25 when i = 1)

**Comment by stevekuznetsov** [comment]
Why should `prow` have to know about this? We auto-scale our nodes and use a DNS solution that works, always, so we never hit this. IMO this is the wrong layer.

File: prow/clonerefs/run.go (hunk 1 above)
Comment (coryrc) [reply prow/clonerefs/run.go:91]: I need to output when it does fail under some scenarios too

File: prow/clonerefs/run.go (hunk 2 above)
Comment (coryrc) [reply prow/clonerefs/run.go:84]: I'm missing the fails-3-times scenario writing-to-channel. Will figure out.

File: prow/clonerefs/run.go (hunk 3 above)
Comment (coryrc) [reply prow/clonerefs/run.go:96]: Why not? That's the intention.

File: prow/clonerefs/run.go (hunk 4 above)
Comment (coryrc) [reply prow/clonerefs/run.go:94]: delay is fine as a name

File: prow/clonerefs/run.go (hunk 3 above)
Comment (coryrc) [reply prow/clonerefs/run.go:96]: oh, I see; I'm doing the sleep wrong. Oh darn.

Comment (fejta) [reply prow/clonerefs/run.go:96]: It failed all four times and will now exit. There's no need to make it sleep for 400 seconds before reporting this failure.

**Comment by coryrc** [comment]
> Why should `prow` have to know about this? We auto-scale our nodes and use a DNS solution that works, always, so we never hit this. IMO this is the wrong layer.

This is the only

**Comment by coryrc** [comment]
(wrong button)

**Comment by coryrc** [comment]
> Why should `prow` have to know about this? We auto-scale our nodes and use a DNS solution that works, always, so we never hit this. IMO this is the wrong layer.

Because we haven't been able to figure it out.

**Comment by fejta** [comment]
The clonerefs image prow injects is controlled here (does not need to be the same as the rest of prow):

https://github.com/kubernetes/test-infra/blob/a2bf5f85155043c7ca2dc01661d1d710a1bb46a6/config/prow/config.yaml#L9-L14

**Comment by alvaroaleman** [comment]
> Why should prow have to know about this? We auto-scale our nodes and use a DNS solution that works, always, so we never hit this. IMO this is the wrong layer

Yeah. Wild guess would be that you use node-local-dns DaemonSet and have nothing that taints your nodes until those containers are up and ready. Please properly debug your issue instead of adding a bandaid-workaround here.

I am not  generally opposed to have retrying as everything might fail intermittently if a network is involved but:
* Please do not make that retrying specific to DNS failures, make it generic with a whitelist of known terminal errors that are not retried (e.G. 404 when cloning)
* Both the initial retry interval and the total time spent retrying are _way_ to high in my opinion

**Comment by alvaroaleman** [comment]
Also, the `fixes` directives on the PR body are misleading. Kubernetes without nodelocal DNS Daemonset has known issues with DNS under load due to a conntrack race (xref https://github.com/kubernetes/kubernetes/issues/56903). That error is not specific to using git, it just happens to be that git doesn't retry DNS lookups.

**Comment by cjwagner** [comment]
If you want to build your own version of clonerefs that includes your workaround there are instructions for publishing the Prow component images here: https://github.com/kubernetes/test-infra/blob/master/prow/build_test_update.md#how-to-update-the-cluster

> Update your deployment (optionally build/pushing the image) to a new image with:
> ```shell
> # export PROW_REPO_OVERRIDE=gcr.io/k8s-prow  # optionally override project
> push.sh  # Build and push the current repo state.
> bump.sh --list  # Choose a recent published version
> bump.sh v20181002-deadbeef # Use a specific version
> ```

**Comment by coryrc** [comment]
Will be using private clonerefs image

**Pull request closed without merging**